mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp

```diff
@@ -73,6 +73,32 @@ static bool isWriteHintOrNone(const CachePolicyAttr &attr) {
          kind == CachePolicy::WRITE_BACK || kind == CachePolicy::WRITE_THROUGH;
 }
 
+// Validations for nd instruction arguments is successful if any of these are
+// true:
+// - tensor descriptor and the output vector shapes exactly match.
+// - tensor descriptor has a sg_map attribute and the distributed vector shape
+//   matches the tensor descriptor shape when scaled using sg_map factors on
+//   each dimension.
+static bool isArgShapesValid(ArrayRef<int64_t> descShape,
+                             ArrayRef<int64_t> valShape, SGMapAttr sgMap) {
+  if (descShape == valShape)
```

Comment:
what if descShape == valShape and sgMap is valid? does it mean the sgMap will be discarded?

Reply:
Sorry, I didn't get the question. If the shapes are equal that means the nd load/store is valid and not distributed (if there's an sg map, only sg map's validity is checked by `SGMapAttr::verify`). If the shapes don't match then we either have a distributed or an invalid case. For distributed we check that ranks are the same, the sg map is present and the scaled values for each dimension match.

Reply:
NVM, it seems a dumb question. It seems to me that, there are currently 3 stages: 1) pure SIMD code, `sgMap == null`, and `descShape == valShape`. 2). a valid sgMap is attached to guide the lowering, but the code is not rewritten yet, so we have `descShape == valShape` and `sgMap != null`, but sgMap is not effective yet. 3). code is rewritten, so `descShape != valShape`, and `sgMap != null`, and sgMap is now effective.

Reply:
Yup, those are all valid states